libgalois/test/reduce-error-info.cpp

```diff
@@ -0,0 +1,103 @@
+#include <algorithm>
+#include <functional>
+#include <iostream>
+
+#include "katana/Galois.h"
+#include "katana/ResultReduction.h"
+
+katana::Result<void>
+ErrorOnTrue(bool error) {
+  if (error) {
+    return KATANA_ERROR(katana::ErrorCode::NotFound, "error");
+  }
+  return katana::ResultSuccess();
+}
+
+katana::Result<bool>
+ErrorOnTrue(bool error, bool return_value) {
+  if (error) {
+    return KATANA_ERROR(katana::ErrorCode::NotFound, "error");
+  }
+  return return_value;
+}
+
+void
+TestVoidFunc() {
+  katana::GReduceCopyableErrorInfo error_info;
```

Comment:
I think we need better names. How about `katana::CombinedErrorInfo`? The fact that it's implemented via `GReducible` or even Reduction is an implementation detail and doesn't serve the API user. 

Similarly, `KATANA_REDUCTION_CHECKED` should be renamed to `KATANA_COMBINE_ERROR_INFO`. Feels clearer to me as to what to expect from the macro.

Reply:
Sounds good to me.

Reply:
Actually, after having thought about it (and seeing your comment about ANY or OR), I've changed my thinking:

1) We should implement both ANY and ALL semantics. Currently, it seems you've implemented ANY, or first thread to register an error (in the thread ID order). In the future, we can support ALL, where we combine errors into a `bit vector` or  some such structure. 
2) We should rename the object to convey the intent of ANY or ALL. In this case, for ANY semantics, it should be `katana::GatherAnyError` or `GatherAtMostOneError`. For ALL case, it should be `katana::GatherAllErrors`. The macro can be similarly named to `KATANA_GATHER_ERROR_INFO`.

Reply:
Done. I was also thinking of renaming ResultReduction.h to CombinedErrorInfo.h but didn't do it to be more consistent with Result.h and Reduction.h

Reply:
I don't think there's a need for both ANY and ALL. We should switch from ANY to ALL in the future. Ideally, it would be ALL-DISTINCT. Having more options can be confusing. IMO the current `CombinedErrorInfo` should be updated later to combine all distinct error infos into one error info. Currently, ANY is sufficient so that we can at least detect and return some error.